docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+        // If the taker can be anybody, then there can be no taker tokens
+        if (offer.taker == address(0)) {
+            // If taker not specified, then can't specify IDs
+            require(offer.takerIds.length == 0, "If trade is offered to anybody, cannot specify tokens from taker.");
+        } else {
+            // Ensure the taker owns the taker tokens
+            for (uint i = 0; i < offer.takerIds.length; i++) {
+                require(idToOwner[offer.takerIds[i]] == offer.taker, "At least one taker token doesn't belong to taker.");
+            }
+        }
+        return true;
+    }
+
+    function cancelOffer(address maker, address taker, uint256 makerWei, uint256[] memory makerIds, uint256 takerWei, uint256[] memory takerIds, uint256 expiry, uint256 salt) external {
+        require(maker == msg.sender, "Only the maker can cancel this offer.");
+        Offer memory offer = Offer(maker, taker, makerWei, makerIds, takerWei, takerIds, expiry, salt);
+        bytes32 hash = hashOffer(offer);
+        cancelledOffers[hash] = true;
+        emit OfferCancelled(hash);
+    }
+
+    function acceptTrade(address maker, address taker, uint256 makerWei, uint256[] memory makerIds, uint256 takerWei, uint256[] memory takerIds, uint256 expiry, uint256 salt, bytes memory signature) external payable reentrancyGuard {
+        require(!marketPaused, "Market is paused.");
+        require(msg.sender != maker, "Can't accept ones own trade.");
+        Offer memory offer = Offer(maker, taker, makerWei, makerIds, takerWei, takerIds, expiry, salt);
+        if (msg.value > 0) {
+            ethBalance[msg.sender] = ethBalance[msg.sender].add(msg.value);
+            emit Deposit(msg.sender, msg.value);
+        }
+        require(offer.taker == address(0) || offer.taker == msg.sender, "Not the recipient of this offer.");
+        require(tradeValid(maker, taker, makerWei, makerIds, takerWei, takerIds, expiry, salt, signature), "Trade not valid.");
+        require(ethBalance[msg.sender] >= offer.takerWei, "Insufficient funds to execute trade.");
+        // Transfer ETH
+        ethBalance[offer.maker] = ethBalance[offer.maker].sub(offer.makerWei);
+        ethBalance[msg.sender] = ethBalance[msg.sender].add(offer.makerWei);
+        ethBalance[msg.sender] = ethBalance[msg.sender].sub(offer.takerWei);
+        ethBalance[offer.maker] = ethBalance[offer.maker].add(offer.takerWei);
+        // Transfer maker ids to taker (msg.sender)
+        for (uint i = 0; i < makerIds.length; i++) {
+            _transfer(msg.sender, makerIds[i]);
+        }
+        // Transfer taker ids to maker
+        for (uint i = 0; i < takerIds.length; i++) {
+            _transfer(maker, takerIds[i]);
+        }
+        // Prevent a replay attack on this offer
+        bytes32 hash = hashOffer(offer);
+        cancelledOffers[hash] = true;
+        emit Trade(hash, offer.maker, msg.sender, offer.makerWei, offer.makerIds, offer.takerWei, offer.takerIds);
+    }
```

Comment:
Accept the offer and resolve payments and transfers. 
Taker of offer may add eth balance in the same call which will trigger a Deposit event
Checks market is not paused. 
Checks the trade is valid
Checks funds for taker
Transfer ether
Transfer NFT
Cancelled the offer so it cannot be replayed
Emit a trade event